pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -119,6 +121,98 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
+// BuildDeltaClusters will generate the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
+// Otherwise, we fall back onto generating everything.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, push *model.PushContext,
+	updates *model.PushRequest, watched *model.WatchedResource) ([]*discovery.Resource, []string, model.XdsLogDetails) {
+	clusters := make([]*cluster.Cluster, 0)
+	resources := model.Resources{}
+	removedClusterNames := make([]string, 0)
+	services := make([]*model.Service, 0)
+	envoyFilterPatches := push.EnvoyFilters(proxy)
+	cb := NewClusterBuilder(proxy, push, configgen.Cache)
+	instances := proxy.ServiceInstances
+	// delta when only services are modified and we attempt deltas
+	// we can progressively optimize this
+	delta := updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.ServiceEntry)
```

Comment:
allConfigKeysOfType - This check is not clear to me. I may be missing some thing. So if PushRequest contains other configs like VS, DR other than service - we do not generate deltas? Any reason??

Reply:
the reason is so we don't have a very large change in one PR. For now, we decided it would be best to start of with services, and incrementally move onto DestinationRule, VirtualService, etc.

We will eventually generate deltas for all, at which point we will not need this check.

Reply:
Got it. So delta would kick in only when updates have all services? In that case I would suggest to add method like 

`shouldUseDelta` and pass in updates to it. and there you check if updates has only service changes. So slowly we can make changes to that method - we can reuse that method for listeners etc. So we do not have to change individual builders

Reply:
@adiprerepa This is missed. But I am ok to do this as follow-up